### PR TITLE
#380 #311 Added roles verification to controller methods

### PR DESCRIPTION
**Pull request by rui-castro**
Issue #380 #311 
Some controllers have methods without RodaUser. Should these methods also checkRoles?
The FIXME comments saying "FIXME check user permissions", should be removed?

### CODE REVIEW

**Comment by luis100** [comment]
In principle, all methods should check roles, but there might be some exceptions. Which methods are lacking the check?

The FIXME comment should be removed.

**Comment by rui-castro** [comment]
@luis100 

- https://github.com/rui-castro/roda/blob/rcastro_review_roles_380/roda-ui/roda-wui/src/main/java/org/roda/wui/api/controllers/Formats.java#L74
- https://github.com/rui-castro/roda/blob/rcastro_review_roles_380/roda-ui/roda-wui/src/main/java/org/roda/wui/api/controllers/Notifications.java#L77
- https://github.com/rui-castro/roda/blob/rcastro_review_roles_380/roda-ui/roda-wui/src/main/java/org/roda/wui/api/controllers/Risks.java#L75
- All methods in Theme https://github.com/rui-castro/roda/blob/rcastro_review_roles_380/roda-ui/roda-wui/src/main/java/org/roda/wui/api/controllers/Theme.java#L28 
- https://github.com/rui-castro/roda/blob/rcastro_review_roles_380/roda-ui/roda-wui/src/main/java/org/roda/wui/api/controllers/UserManagement.java#L202
- https://github.com/rui-castro/roda/blob/rcastro_review_roles_380/roda-ui/roda-wui/src/main/java/org/roda/wui/api/controllers/Theme.java#L28

**Comment by rui-castro** [comment]
@luis100 FIXME comments removed

**Comment by luis100** [comment]
The only exception to the verification are the Theme, all others must check the roles.

**Comment by hsilva-keep** [comment]
@rui-castro , after analysing your commits, some comments before we can integrate your changes:
* Try not to change the imports order. It makes it harder to do a code comparison and we're polluting the history with unnecessary changes.
* When logging, do not concatenate strings as slf4j offers methods with formatters. E.g.:
```java
LOGGER.info("Created LDAP role " + role);
```
vs.
```java
LOGGER.info("Created LDAP role {}", role);
```
* Use eclipse code formatter and configure it to use RODA formatter (https://github.com/keeps/roda/blob/master/code-style/eclipse_formatter.xml)
* Even if debug logging is activated, don't log twice (code method and invoking method). E.g. RodaCoreFactory.createRoles() and LdapUtility.addRole(). Also, inside a catch with expected behaviour, just warn about what happened and don't pass it the exception.

**Comment by luis100** [comment]
@rui-castro please fix both above comments and notify

File: roda-ui/roda-wui/src/main/java/org/roda/wui/api/controllers/UserManagement.java [hunk 1]
```diff
@@ -323,6 +321,10 @@ public static void removeGroup(RodaUser user, String groupname)
     registerAction(user, "UserManagement", "removeGroup", null, duration, "groupname", groupname);
   }
 
+  // TODO: Methods bellow this line should also checkRoles? If so, a RodaUser is needed.
```
Comment (rui-castro) [inline roda-ui/roda-wui/src/main/java/org/roda/wui/api/controllers/UserManagement.java]: @luis100 @hsilva-keep The methods bellow this comment, are all related with user registration/recover and don't have access to the RodaUser that is calling the method. To call chechRoles(...) the RodaUser must come from somewhere. Please advise on what to do.

**Comment by rui-castro** [comment]
@luis100 @hsilva-keep Please review the new commits, comment above and the following...
This method UserLogin#login(...) (https://github.com/keeps/roda/blob/master/roda-ui/roda-wui/src/main/java/org/roda/wui/api/controllers/UserLogin.java#L26) should also checkRoles? Does that make sense?

**Comment by hsilva-keep** [comment]
Rui, me & @luis100 have analysed your solution for having a generic way of checking if a user has the required roles for executing a particular method (by creating a new Object inside each method). And realized that if we really need to create that inner class for each method to be able to know the class and method name, then we could/should use that object for useful stuff (like registering the action, calculating the duration, etc.), and we created what we've called [ControllerAssistant.java](https://github.com/keeps/roda/blob/hsilva_dev/roda-ui/roda-wui/src/main/java/org/roda/wui/common/ControllerAssistant.java) So, instead of having
```java
public static Risk createRisk(RodaUser user, Risk risk)
    throws AuthorizationDeniedException, RequestNotValidException, NotFoundException, GenericException {
    Date startDate = new Date();

    // check user permissions
    UserUtility.checkRoles(user, new Object(){}.getClass().getEnclosingMethod());

    RodaCoreFactory.getModelService().createRisk(risk, false);

    // register action
    long duration = new Date().getTime() - startDate.getTime();
    registerAction(user, RISKS_COMPONENT, "createRisk", null, duration, "risk", risk);

    return risk;
  }
```
we can have
```java
public static Risk createRisk(RodaUser user, Risk risk)
    throws AuthorizationDeniedException, RequestNotValidException, NotFoundException, GenericException {
    ControllerAssistant controllerAssistant = new ControllerAssistant() {};

    controllerAssistant.checkRoles(user);

    RodaCoreFactory.getModelService().createRisk(risk, false);

    controllerAssistant.registerAction(user, "risk", risk);

    return risk;
  }
```

which does the same but in a cleaner way.

Can you change the code for doing using this new/leaner way to check roles and register action in all controllers?

**Comment by rui-castro** [comment]
@hsilva-keep @luis100 ok. [ControllerAssistant.java](https://github.com/keeps/roda/blob/hsilva_dev/roda-ui/roda-wui/src/main/java/org/roda/wui/common/ControllerAssistant.java) is in branch `hsilva_dev`, are you gonna merge into master or should I rebase my branch in branch `hsilva_dev`?

**Comment by hsilva-keep** [comment]
Well spotted @rui-castro . I've merged master and now it points to the commit that has the new class.

**Comment by hsilva-keep** [comment]
@rui-castro I've updated RODA master branch as my previous commit introduced an NPE due to method overload. Please sync your repo.

**Comment by rui-castro** [comment]
@hsilva-keep @luis100 done.

**Comment by hsilva-keep** [comment]
Already merged.

**Pull request closed without merging**